File: app/src/main/java/com/azure/minitwitter/data/TweetRepository.java
package com.azure.minitwitter.data;

import android.arch.lifecycle.MutableLiveData;
import android.widget.Toast;

import com.azure.minitwitter.common.Constants;
import com.azure.minitwitter.common.MyApp;
import com.azure.minitwitter.common.SharedPreferencesManager;
import com.azure.minitwitter.retrofit.AuthTwitterService;
import com.azure.minitwitter.retrofit.request.RequestCreateTweet;
import com.azure.minitwitter.retrofit.response.AuthTwitterClient;
import com.azure.minitwitter.retrofit.response.Like;
import com.azure.minitwitter.retrofit.response.Tweet;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class TweetRepository {

    AuthTwitterService authTwitterService;
    AuthTwitterClient authTwitterClient;
    MutableLiveData<List<Tweet>> allTweets;
    MutableLiveData<List<Tweet>> favTweets;
    String userName;

    public TweetRepository(){
        authTwitterClient = AuthTwitterClient.getInstance();
        authTwitterService = authTwitterClient.getAuthTwitterService();
        allTweets = getAllTweets();
        userName = SharedPreferencesManager.getSomeStringValue(Constants.PREF_USERNAME);
    }

    public MutableLiveData<List<Tweet>> getAllTweets() {

        if(allTweets == null){
            allTweets = new MutableLiveData<>();
        }

        Call<List<Tweet>> call = authTwitterService.getAllTweets();
        call.enqueue(new Callback<List<Tweet>>() {
            @Override
            public void onResponse(Call<List<Tweet>> call, Response<List<Tweet>> response) {
                if(response.isSuccessful()){
                    allTweets.setValue(response.body());
                }else{
                    Toast.makeText(MyApp.getContext(),"Something goes wrong", Toast.LENGTH_SHORT);
                }
            }

            @Override
            public void onFailure(Call<List<Tweet>> call, Throwable t) {
                Toast.makeText(MyApp.getContext(), "Error on connection", Toast.LENGTH_SHORT);
            }
        });

        return allTweets;
    }

    public MutableLiveData<List<Tweet>> getFavTweets() {
        if(favTweets == null){
            favTweets = new MutableLiveData<>();
        }

        List<Tweet> newFavList = new ArrayList<>();
        Iterator itTweets = allTweets.getValue().iterator();

        while(itTweets.hasNext()){
            Tweet current = (Tweet) itTweets.next();
            Iterator itLikes = current.getLikes().iterator();
            boolean found = false;

            while(itLikes.hasNext() && !found){
                Like like = (Like)itLikes.next();

                if(like.getUsername().equals(userName)){
                    found = true;
                    newFavList.add(current);
                }
            }
        }

        favTweets.setValue(newFavList);

        return favTweets;
    }

        public void createTweet(String message){
        RequestCreateTweet requestCreateTweet = new RequestCreateTweet(message);
        Call<Tweet> call = authTwitterService.createTweet(requestCreateTweet);

        call.enqueue(new Callback<Tweet>() {
            @Override
            public void onResponse(Call<Tweet> call, Response<Tweet> response) {
                if(response.isSuccessful()){
                    List<Tweet> listCloned = new ArrayList<>();
                    listCloned.add(response.body());

                    for(Tweet tweet: allTweets.getValue()){
                        listCloned.add(new Tweet(tweet));
                    }

                    allTweets.setValue(listCloned);
                } else {
                    // TODO: Refactor to include the error message on string
                    Toast.makeText(MyApp.getContext(), "Something goes wrong",Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Tweet> call, Throwable t) {
                // TODO: Refactor to include the error message on string
                Toast.makeText(MyApp.getContext(), "Error on connection",Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void likeTweet(final int idTweet){
        Call<Tweet> call = authTwitterService.likeTweet(idTweet);

        call.enqueue(new Callback<Tweet>() {
            @Override
            public void onResponse(Call<Tweet> call, Response<Tweet> response) {
                if(response.isSuccessful()){
                    List<Tweet> listCloned = new ArrayList<>();
                    listCloned.add(response.body());

                    for(Tweet tweet: allTweets.getValue()){
                        Tweet tweetToInsertOnList;

                        if(tweet.getId() == idTweet){
                            tweetToInsertOnList = response.body();
                        } else {
                            tweetToInsertOnList = new Tweet(tweet);
                        }

                        listCloned.add(tweetToInsertOnList);
                    }

                    allTweets.setValue(listCloned);
                    getFavTweets();
                } else {
                    // TODO: Refactor to include the error message on string
                    Toast.makeText(MyApp.getContext(), "Something goes wrong",Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Tweet> call, Throwable t) {
                // TODO: Refactor to include the error message on string
                Toast.makeText(MyApp.getContext(), "Error on connection",Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/azure/minitwitter/data/TweetRepository.kt
package com.azure.minitwitter.data

import android.arch.lifecycle.MutableLiveData
import android.widget.Toast
import com.azure.minitwitter.common.Constants
import com.azure.minitwitter.common.MyApp
import com.azure.minitwitter.common.SharedPreferencesManager
import com.azure.minitwitter.retrofit.AuthTwitterService
import com.azure.minitwitter.retrofit.request.RequestCreateTweet
import com.azure.minitwitter.retrofit.response.AuthTwitterClient
import com.azure.minitwitter.retrofit.response.Tweet
import com.azure.minitwitter.retrofit.response.TweetDeleted
import retrofit2.Call
import retrofit2.Callback;
import retrofit2.Response

class TweetRepository{

    val authTwitterClient: AuthTwitterClient = AuthTwitterClient.instance
    val authTwitterService: AuthTwitterService = authTwitterClient.authTwitterService
    private var allTweets: MutableLiveData<List<Tweet>> = getAllTweets()
    private lateinit var favTweets: MutableLiveData<List<Tweet>>
    val username = SharedPreferencesManager.getSomeStringValue(Constants.PREF_USERNAME)

    fun getAllTweets(): MutableLiveData<List<Tweet>>{

        if(allTweets == null){
            allTweets = MutableLiveData()
        }

        val call = authTwitterService.getAllTweets()
        call.enqueue(object : Callback<List<Tweet>>{
            override fun onResponse(call: Call<List<Tweet>>, response: Response<List<Tweet>>) {
                if(response.isSuccessful){
                    allTweets.setValue(response.body())
                } else {
                    Toast.makeText(MyApp.context, "Something goes wrong", Toast.LENGTH_SHORT)
                }
            }

            override fun onFailure(call: Call<List<Tweet>>, t: Throwable) {
                Toast.makeText(MyApp.context, "Error on connection", Toast.LENGTH_SHORT)
            }
        })

        return allTweets
    }

    fun getFavTweets(): MutableLiveData<List<Tweet>>{

        if(favTweets == null){
            favTweets = MutableLiveData()
        }

        val newFavList = ArrayList<Tweet>()
        val iterator = allTweets.value!!.iterator()

        while(iterator.hasNext()){
            val current = iterator.next()
            val itLikes = current.likes.iterator()
            var found = false

            while(itLikes.hasNext() && !found){
                val like = itLikes.next()

                if(like.username == username){
                    found = true
                    newFavList.add(current)
                }
            }
        }

        favTweets.value = newFavList
        return favTweets
    }

    fun createTweet(message: String){
        val requestCreateTweet = RequestCreateTweet(message)
        val call = authTwitterService.createTweet(requestCreateTweet)

        call.enqueue(object : Callback<Tweet>{
            override fun onResponse(call: Call<Tweet>, response: Response<Tweet>) {
                if(response.isSuccessful){
                    val listClonned = ArrayList<Tweet>()
                    listClonned.add(response.body()!!)

                    allTweets.value?.forEach { tweet: Tweet ->
                        listClonned.add(Tweet(tweet))
                    }

                    allTweets.value = listClonned
                } else {
                    // TODO: Refactor to include the error message on string
                    Toast.makeText(MyApp.context, "Something goes wrong", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Tweet>, t: Throwable) {
                // TODO: Refactor to include the error message on string
                Toast.makeText(MyApp.context, "Error on connection", Toast.LENGTH_SHORT).show()
            }
        })
    }

    fun likeTweet(idTweet: Int){
        val call = authTwitterService.likeTweet(idTweet)

        call.enqueue(object : Callback<Tweet>{
            override fun onResponse(call: Call<Tweet>, response: Response<Tweet>) {
                if(response.isSuccessful){
                    val listClonned = ArrayList<Tweet>()
                    listClonned.add(response.body()!!)

                    allTweets.value?.forEach {tweet: Tweet ->
                        val tweetToInsertOnList =
                                if(tweet.id == idTweet){
                                    response.body()
                                } else {
                                    Tweet(tweet)
                                }

                        listClonned.add(tweetToInsertOnList!!)
                    }

                    allTweets.value = listClonned
                    getFavTweets()
                } else {
                    // TODO: Refactor to include the error message on string
                    Toast.makeText(MyApp.context, "Something goes wrong", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Tweet>, t: Throwable) {
                // TODO: Refactor to include the error message on string
                Toast.makeText(MyApp.context, "Error on connection", Toast.LENGTH_SHORT).show()
            }
        })
    }

    fun deleteTweet(idTweet: Int){
        val call = authTwitterService.deleteTweet(idTweet)

        call.enqueue(object : Callback<TweetDeleted>{
            override fun onResponse(call: Call<TweetDeleted>, response: Response<TweetDeleted>) {
                if(response.isSuccessful){
                    val clonnedTweets = ArrayList<Tweet>()

                    allTweets.value?.forEach { tweet: Tweet ->
                        if(tweet.id != idTweet){
                            clonnedTweets.add(Tweet(tweet))
                        }
                    }

                    allTweets.value = clonnedTweets
                    getFavTweets()
                } else {
                    // TODO: Refactor to include the error message on string
                    Toast.makeText(MyApp.context, "Something goes wrong", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<TweetDeleted>, t: Throwable) {
                // TODO: Refactor to include the error message on string
                Toast.makeText(MyApp.context, "Error on connection", Toast.LENGTH_SHORT).show()
            }
        })
    }

}